### ISSUE TITLE
Tedlium perplex on original eaxct vocab and test set from kaldi recipe

### ISSUE THREAD

**Issue by vince62s** [danpovey/pocolm#13]
Let me know if this is what you expected: 

https://github.com/vince62s/pocolm/blob/perplex/egs/tedlium/perplex.sh

**Comment by danpovey** [comment]
Yes, that looks great.  Can you make a pull request as e.g.
egs/tedlium/local/srilm_baseline_pruning.sh
and make it a script that can run, with perplexities in comments?

**Comment by vince62s** [comment]
As a matter of fact I adjusted the srilm_baseline.sh to make it a generic script to put in /scripts.
Also I changed some logics by namingthe training text always data/text/train.txt as you named data/text/dev.txt and the vocab becomes vocab.txt without the size in the name.
It makes the run.sh much easier and generic when copying, same for srilm_baseline.

**Comment by danpovey** [comment]
> As a matter of fact I adjusted the srilm_baseline.sh to make it a generic
> script to put in /scripts.
> Also I changed some logics by naming the training text always
> data/text/train.txt

Naming the training text train.txt is a little problematic because part of
the concept of the toolkit is to naturally handle the case where there are
multiple sources of training data.  In this case they all have different
names and there is no "special" one.  It's not clear to me how your scripts
would extend to the setup in the swbd_fisher example, where there are two
different training sets.

Can you try to create a script that's just in local/ for now?

Dan

**Comment by danpovey** [comment]
... also bear in mind that SRILM is not really considered part of pocolm.
That is, pocolm has no essential dependency on SRILM.  Putting a script
that depends on srilm in scripts/ makes this unclear.

Dan

**Comment by vince62s** [comment]
okay I see. I'll have an issue with the dev set, I don't want to build some complex stuff just for the 146kB file ... might be easier to just post it.

FYI maxent is in between poco and kn. For order 4: 176.91 / 179.83 / 185.99

**Comment by danpovey** [comment]
Can you explain what the issue is, regarding the dev set?  and "build some
complex stuff"?  I don't understand.

What I wanted is for you to create a pull request to add a script in local/
containing your SRILM pruning commands, just so we have conveniently
accessible documentation of how to do this pruning in SRILM.

**Comment by vince62s** [comment]
If you recall, all my perplexity calculation are based on the dev set from the TEDLIUM text extraction of the test set from the corpus. This is not the first n sentences from the cantab text corpus.
hence I have this dev set on disk, but no script to extract it from the TEDLIUM release1/2.
Follow me ?

**Comment by danpovey** [comment]
Oh.
Perhaps you could write a script to do a similar experiment but use the
first n sentences of the cantab corpus as the test set, instead?  That will
also be easier when adapting the same script to other setups such as
Switchboard.
Dan

**Comment by vince62s** [comment]
here we go. Using a less pruned at decoding time (best results), then the impact of Poco versus original Cantab order 4 LM for rescoring is as follow :

Avec LM decoding = small_arpa_lm=db/lm.o3g.pruned2e8.gz = 97.6 MB - OK
WER 11.5 | 507 17792 | 90.4 7.1 2.6 1.9 11.5 81.1 | -0.057 | exp/nnet2_online/nnet_ms_sp_smbr_0.000005/decode_epoch4_dev_vn/score_10_0.0/ctm.filt.filt.sys
WER 10.8 | 507 17792 | 91.0 6.7 2.3 1.8 10.8 79.9 | -0.164 | exp/nnet2_online/nnet_ms_sp_smbr_0.000005/decode_epoch4_dev_vn.rescore/score_10_0.0/ctm.filt.filt.sys
WER 11.0 | 1155 27512 | 90.5 6.7 2.8 1.5 11.0 76.9 | -0.068 | exp/nnet2_online/nnet_ms_sp_smbr_0.000005/decode_epoch4_test_vn/score_10_0.0/ctm.filt.filt.sys
WER 10.4 | 1155 27512 | 91.0 6.0 3.0 1.4 10.4 74.5 | -0.171 | exp/nnet2_online/nnet_ms_sp_smbr_0.000005/decode_epoch4_test_vn.rescore/score_11_0.0/ctm.filt.filt.sys

rescoring with Poco order 4 LM - big_arpa_lm=db/150000_4gram_unpruned.arpa.gz
WER 10.6 | 507 17792 | 91.2 6.5 2.3 1.8 10.6 80.1 | -0.160 | exp/nnet2_online/nnet_ms_sp_smbr_0.000005/decode_epoch4_dev_vn.rescore/score_10_0.0/ctm.filt.filt.sys
WER 10.3 | 1155 27512 | 91.2 6.1 2.7 1.5 10.3 73.7 | -0.134 | exp/nnet2_online/nnet_ms_sp_smbr_0.000005/decode_epoch4_test_vn.rescore/score_10_0.0/ctm.filt.filt.sys

**Comment by danpovey** [comment]
Cool!  So it is a little better, 10.8 -> 10.6 and 10.4 -> 10.3.  That is
about what I expected, because for the unpruned LM, the perplexity
differences are not that great (e.g. 78 vs 80).
I just got some plots from Ke and Zhouyang, and the pocolm pruning is very
much more effective than the SRILM pruning.  (Actually this was true of
kaldi_lm too, but the toolkit was harder to use and didn't support LM
interpolation very well).
So I suspect you'll see more improvement if you do the first-pass decoding
with pocolm-pruned language models.

Dan

**Comment by vince62s** [comment]
So let me know when I can convert pruned to arpa...

**Comment by danpovey** [comment]
It's already done.

**Comment by vince62s** [comment]
I'll send you the full results by email, too much data for here.
